### ISSUE TITLE
Improvement JS Management

### ISSUE THREAD

**Issue by Volker-E** [owncloud/core#15400]
This is a general collection issue with current state of handling and including JS files in ownCloud.
The [demo (ownCloud 8)](https://demo.owncloud.org/) uses in the /apps/files/ view **73 JS files with (partly) unminified 2.2 MB of data**. 

From a front-end dev point of view with 2015's best practices in mind, I'm convinced that there's a necessity to critically question the inclusion and handling of JS functionality/files right now for performance and devices support's regard.

Main target IMHO should be a decrease of files needed, better output handling, maybe depending on what view the user is on.

**Comment by LukasReschke** [comment]
https://github.com/owncloud/core/blob/4996706078a901fa532b0fed0fb0bc264d1ba778/config/config.sample.php#L870-L874

**Comment by Volker-E** [comment]
Yeah, that's an important step at the end of performance improvement cycle, but the sheer amount of 2.2MB of JS-only data is too much.

First view is very slow, in [this test at 7.8s](http://www.webpagetest.org/result/150404_8Y_NKJ/). And lots of that originates in JS handling.

Moreover concatenation isn't an answer to everything. There's a limit to when the files concatenation to one file is outperformed by concatenation to a few files. See also http://stackoverflow.com/questions/9048721/max-size-of-minify-combining-js-files.
My pledge is for somebody responsible/policies taking care of (excuse my missing information of possible people/policies)
* JS features included in ownCloud,
* separation of 3rd-party libraries (assets/vendor directory) from ownCloud's own JS files.
* JS best practices & coding guidelines, that should be followed like f.e. performance of usage of jQuery selectors versus native implementations  etc.
* maybe including an JS dependency loader 

Some of the concerns here are already discussed at #11973

**Comment by MorrisJobke** [comment]
We also have a lot of duplicated libraries. But the migration from three different version of a advanced select box handling library to a single one took some time to do. Now we simply use them in parallel (jquery.multiselect.js, multiselect.js and choosen) ... this is just one example. :(

**Comment by PVince81** [comment]
Some JS files are included even when not used at all, simply because it's not possible for apps to detect whether the current page is the files app or login page. So they **always** include the script by putting them in "appinfo/app.php". There is currently no other way.
@nickvergessen and I had some discussion about this: https://github.com/owncloud/core/pull/15073

**Comment by nickvergessen** [comment]
@PVince81 well there are other ways, althought they may have point of failures and are not really what it should be like: https://github.com/owncloud/calendar/blob/master/appinfo/app.php#L42-L45

**Comment by Volker-E** [comment]
@MorrisJobke @PVince81 @nickvergessen All valid points (of failure) and not uncommon to projects, that have been growing over time. 
My conviction is, that the current state of front-end performance is far from good and that there should be guidelines for developers on how to handle/integrate JS functionality *the right way* and to collect and identify current shortcomings here, like the mentioned three-way integration of select box handling lib and referencing them here.

Concerning differentiation of views like login vs files app. Has there probably been a discussion on adding handlers similar to [conditional tags in WordPress](https://codex.wordpress.org/Function_Reference/is_page)? Another thing that I'm missing for styling purposes are body classes like `view--login`, `view--files`. But that's a different issue.

**Comment by PVince81** [comment]
@Volker-E the problem is that guidelines won't help with the JS problem of apps. The problem is that there is currently no other way of doing it. It's an architecture problem. We need to improve it by providing ways for app to register their stuff more semantically and on-demand instead of always.

We thought about using frameworks like requirejs. The problem is that if we put it in core it will require every app to do so. But some apps are written in angular or something else, so core shouldn't have too strong requirements about the way to load JS files, but maybe provide the tools and let app devs choose.

Also, most of the JS code in core isn't even good JS... it's just code splattered all over the place which isn't even unit testable. It will take some time to make core ready for whichever better JS loading mechanism will be used at some point.

**Comment by Volker-E** [comment]
@PVince81 Thanks for the response. I completely get your points, particularly about freedom of choice for developers.

In my experience especially with Free/Libre Open Source projects, but in general with every bigger software development project with more than 2 people working on. If you have guidelines and best practices you provide for all newcomers and for yourself a single-goto-reference. That must not be set in stone and can/should be improved over time.
But there is a reference that helps orientation and faster development.

If there are none at all, IMHO, the project is maybe more attractive to newcomers, but as bug reporter or core maintainer you *can't* criticize on others code, because of an everything goes mentality and not only will the external apps have a huge (to me horrible) diversity of approaches far from being innovative, but also the core gets bloated and less maintainable and motivated (high-level) developers will turn away.
Again, inclusion of three different custom select-box JS-libs is a great example.

Require.js was just a possibility. Let's brainstorm if there are other, better fitting solutions.

**Comment by PVince81** [comment]
There are already guidelines, examples in the doc: https://doc.owncloud.org/server/8.0/developer_manual/app/js.html and the more general app dev docs: https://doc.owncloud.org/server/8.0/developer_manual/app/index.html

But it misses the specific case where an app wants to integrate code into other apps (like for example registering file actions or adding stuff to the login page)

**Comment by Volker-E** [comment]
Nice. 
Is everybody here aware of why it's performance-wise a big problem to have lots of HTTP-Requests and just slightly used JS/CSS files and lots of (maybe unnecessary) DOM interaction included in a web application? Or should I give food/links for thought?

**Comment by oparoz** [comment]
@PVince81 - RequireJS can be used with Angular http://www.sitepoint.com/using-requirejs-angularjs-applications/ and even with ReactJS. I think at some point every project has to decide whether usability is more important than developers' total freedom. Of course, you don't want to alienate your community, but when I was told to include all my js libs in my apps instead of re-using the ones in core and saw that there was no way to manage dependencies in core, I thought it was a recipe for disaster.

@Volker-E - Maybe a documentation PR would be useful, but then it will put the spot on a flaw in core with no solution in sight.

**Comment by LukeOwncloud** [comment]
@Volker-E Thanks for bringing up this issue. I am surprised that there aren't more issues on this topic. Whenever I speak to friends about owncloud their first reaction is: "Oh yes, great project but it (especially the web interface) is too slow for a raspberry pi". I am running it on an Intel core duo. Not the most current hardware but I still expect more speed.

I just tried 8.0.4. Login page: 67 requests. 1.4MB. About 1.6 seconds on my server.
File view: 148 requests. 2.3MB. 1-2 seconds.
What is even worse: I see no caching. All 200 responses. (Not due to this issue: https://github.com/owncloud/core/issues/13219, ie. deflate is disabled)

@LukasReschke I tried enabled pipelining. Horrible result: It takes 5-10 seconds before I even get a response from the server (it is probably busy concatenating all those files).

PS: I am writing this to emphasize the importance of this issue. I don't see an easy solution though, either. :(

**Comment by MorrisJobke** [comment]
> @LukasReschke I tried enabled pipelining. Horrible result: It takes 5-10 seconds before I even get a response from the server (it is probably busy concatenating all those files).

Try the next refresh ;) This is only once per page

**Comment by PVince81** [comment]
I'm also worried by the increasing number of JS files, especially with all the recent new features: collaborative tags, comments, etc.

**Comment by PVince81** [comment]
closing in favor of older ticket https://github.com/owncloud/core/issues/12743
